src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -15,44 +18,90 @@ namespace AutoRest.CSharp.Output.Models.Types
     internal sealed class ModelTypeProvider : TypeProvider
     {
         private readonly TypeFactory _typeFactory;
+        private readonly InputModelType _inputModel;
 
         protected override string DefaultName { get; }
         protected override string DefaultAccessibility { get; }
 
         public IReadOnlyList<FieldDeclaration> Fields { get; }
         public ConstructorSignature PublicConstructor { get; }
+        public ConstructorSignature SerializationConstructor { get; }
+
+        private readonly IReadOnlyDictionary<InputModelProperty, FieldDeclaration> _inputsToFields;
+        private readonly IReadOnlyDictionary<Parameter, FieldDeclaration> _parametersToFields;
 
         public ModelTypeProvider(InputModelType inputModel, TypeFactory typeFactory, string defaultNamespace, SourceInputModel? sourceInputModel)
             : base(inputModel.Namespace ?? defaultNamespace, sourceInputModel)
         {
+            _inputModel = inputModel;
             _typeFactory = typeFactory;
 
             DefaultName = inputModel.Name;
             DefaultAccessibility = inputModel.Accessibility ?? "public";
-            Fields = CreateFields(inputModel).ToArray();
-            PublicConstructor = BuildPublicConstructor(inputModel);
+
+            (_inputsToFields, _parametersToFields) = CreateParametersAndFieldsForRoundTripModel(inputModel, typeFactory);
+
+            Fields = _inputsToFields.Values.ToList();
+            PublicConstructor = BuildPublicConstructor(Declaration.Name, _parametersToFields.Keys.ToList());
+
+            // Since we consider all models roundtrip for now, use the same constructor for serialization
+            SerializationConstructor = PublicConstructor;
```

Comment:
Will this be overridden by derived classes of this type like in a specific type of model?
In mgmt plane, models are initialized using their internal constructor which contains all it properties as parameters

Reply:
No, there is no plan to have types derived from `ModelTypeProvider` yet. `EnumTypeProvider` would be a separate type.